### PR TITLE
process walker perfs: optimize readLimits and readStats

### PR DESCRIPTION
**Pull request by alban**
- optimize readStats
- optimize readLimits
- cache limits and cmdline/name after parsing

Fixes https://github.com/weaveworks/scope/issues/2461

### CODE REVIEW

File: probe/process/walker_linux.go [hunk 1]
```diff
@@ -32,53 +61,107 @@ func readStats(path string) (ppid, threads int, jiffies, rss, rssLimit uint64, e
...
-	rssPages, err = strconv.ParseUint(splits[23], 10, 64)
-	if err != nil {
-		return
+	for spaceCount := 0; pos < len(buf) && spaceCount < procStatFieldThreads-procStatFieldSysJiffies; pos++ {
+		if buf[pos] == ' ' {
+			spaceCount++
+		}
+	}
+	for ; pos < len(buf) && buf[pos] != ' '; pos++ {
+		threads = threads*10 + int(buf[pos]-'0')
```
Comment (2opremio) [inline probe/process/walker_linux.go]: This pattern is repeated very often, please factor this out in a function

**Comment by 2opremio** [comment]
@alban I don't doubt that the code is much faster now but it's very hard to read. Please try to refactor it. I think that separating the parsing of integers from the space-counting code would already be an improvement.

Also, are you sure that parsing ints manually provides an improvement? I wouldn't be surprised if the Golang library wasn't optimized in this respect.

**Comment by alban** [comment]
The issue with [strconv.Atoi](https://golang.org/pkg/strconv/#Atoi) or [strconv.ParseInt](https://golang.org/pkg/strconv/#ParseInt) is that they don't stop parsing when finding a non-digit such as a space and they return an error.

The previous code was using [strings.Fields](https://golang.org/pkg/strings/#Fields) to allocate and build a string compatible with [strconv.Atoi](https://golang.org/pkg/strconv/#Atoi) (containing only the digits without extra spaces). The pprof graphs showed that it was [strings.Fields](https://golang.org/pkg/strings/#Fields) taking the cpu rather than [strconv.Atoi](https://golang.org/pkg/strconv/#Atoi) but we cannot use Atoi without Fields.

`fmt.Sscanf("%d "...)` is able to work with extra spaces but it's also slower.

I'll try to refactor this.

**Comment by alban** [comment]
Code refactored and rebased.

Now that #2456 is merged in master, I ran performance tests again to see if this PR was still bringing better perfs. I used the test scenario described in https://github.com/weaveworks/scope/issues/2461#issuecomment-297036674

## Master branch

Up to 95660e3da735634d566325a43ab33c4380b8d346 (so #2456 is included)

pprof-branch-master

## This PR

pprof-branch-2491

## Conclusion

The function `github.com/weaveworks/scope/probe/process.(*walker).Walk` takes respectively:
- 15.32s
- 12.91s (16% faster) 

Overall, the graphs' titles say respectively:
- 36.37s
- 31.32s (19% faster overall, probably coincidental)

File: probe/process/walker_linux.go [hunk 2]
```diff
@@ -32,53 +89,86 @@ func readStats(path string) (ppid, threads int, jiffies, rss, rssLimit uint64, e
...
+		return 0, nil
+	}
+	pos += len(delim)
+
+	for pos < len(content) && content[pos] == ' ' {
+		pos++
+	}
+
+	var softLimit uint64
+	for ; pos < len(content) && buf[pos] != ' '; pos++ {
```
Comment (2opremio) [inline probe/process/walker_linux.go]: Can't you use `parseUint64WithSpaces()` here too?

File: probe/process/walker_linux.go [hunk 3]
```diff
@@ -18,12 +20,67 @@ type walker struct {
...
+	for spaceCount := 0; *pos < len(*buf) && spaceCount < nSpaces; *pos++ {
+		if (*buf)[*pos] == ' ' {
+			spaceCount++
+		}
+	}
+}
+
+// parseIntWithSpaces is similar to strconv.ParseInt but stops parsing when
+// reading a space instead of returning an error
+func parseIntWithSpaces(buf *[]byte, pos *int) (ret int) {
```
Comment (2opremio) [inline probe/process/walker_linux.go]: Can't you simply use parseUint64WithSpaces and cast to int instead?

Comment (alban) [inline probe/process/walker_linux.go]: Updated.

File: probe/process/walker_linux.go (hunk 2 above)
Comment (alban) [inline probe/process/walker_linux.go]: Updated.

**Pull request merged**